### PR TITLE
LPS-123084 Obtain user segments only when it is needed

### PR DESCRIPTION
**Pull request by cgoncas**
**Motivation**

The main goal of this pull is to call SegmentEntryRetrievers instead read SegmentsWebKeys.SEGMENTS_ENTRY_IDS from the httpRequest (so we can avoid fill SegmentsWebKeys.SEGMENTS_ENTRY_IDS in every request).

This is the third pull to merge into master the Performance Issue fix with Segmentation (for the full solution take a look to #505)

**Proposed solution**

Obtain the segments only where it is used (asset-publisher and collections).

**How to test it**

All manual/funtional tests of Segments and Experiences should work as they did before.

/cc @liferay-echo

### CODE REVIEW

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:tango

**Comment by cgoncas** [comment]
ci:test:echo

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:tango

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by darquesdev** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java [hunk 1]
```diff
@@ -1108,6 +1112,25 @@ protected void activate(Map<String, Object> properties)
...
 
+	private long[] _getSegmentsEntryIds(PermissionChecker permissionChecker) {
+		ServiceContext serviceContext =
+			ServiceContextThreadLocal.getServiceContext();
+
+		if (serviceContext == null) {
+			return new long[0];
+		}
+
+		HttpServletRequest httpServletRequest = serviceContext.getRequest();
```
Comment (darquesdev) [inline modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java:1123]: Would it be better to extract the `HttpServletRequest` from the portlet request? Something like

````
_portal.getHttpServletRequest(portletRequest);

```

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java [hunk 2]
```diff
@@ -2130,8 +2136,10 @@ private String _getSegmentsAnonymousUserId() {
 	}
 
 	private long[] _getSegmentsEntryIds() {
-		return GetterUtil.getLongValues(
-			_portletRequest.getAttribute(SegmentsWebKeys.SEGMENTS_ENTRY_IDS));
+		return _segmentsEntryRetriever.getSegmentsEntryIds(
+			_themeDisplay.getScopeGroupId(), _themeDisplay.getUserId(),
+			_requestContextMapper.map(
+				PortalUtil.getHttpServletRequest(_portletRequest)));
```
Comment (darquesdev) [inline modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/display/context/AssetPublisherDisplayContext.java:2142]: I see a lot of uses of `PortalUtil` in the class, I guess It's ok to use one more time

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java (hunk 1 above)
Comment (cgoncas) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java:1123]: Good catch, thank you :)

File: modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java [hunk 3]
```diff
@@ -109,82 +106,60 @@ protected void doRun(
...
-			return;
-		}
-
-		long[] segmentsEntryIds = _getSegmentsEntryIds(
-			httpServletRequest, themeDisplay.getScopeGroupId(),
-			themeDisplay.getUserId());
-
-		httpServletRequest.setAttribute(
-			SegmentsWebKeys.SEGMENTS_ENTRY_IDS, segmentsEntryIds);
+		if ((layout == null) || layout.isTypeControlPanel() ||
```
Comment (darquesdev) [inline modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java:109]: Nice!! Just one use case left

**Review by darquesdev — APPROVED** [review]
LGTM, just a couple of small changes. Good job! Almost all use cases covered :)

File: modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java (hunk 1 above)
Comment (cgoncas) [reply modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/util/AssetPublisherHelperImpl.java]: Change it!!

File: modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java (hunk 3 above)
Comment (cgoncas) [reply modules/apps/segments/segments-service/src/main/java/com/liferay/segments/internal/events/SegmentsServicePreAction.java]: Thanks!! done :)

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:test:sf

**Comment by cgoncas** [comment]
ci:test:relevant

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:stop

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:forward

**Comment by cgoncas** [comment]
ci:forwar

**Comment by cgoncas** [comment]
ci:forward

**Pull request closed without merging**